In [1]:
!pip install Dbias

  Using cached Dbias-0.1.5-py3-none-any.whl (7.8 kB)
  Using cached numpy-1.26.3-cp39-cp39-win_amd64.whl.metadata (61 kB)
  Using cached spacy-3.7.2-cp39-cp39-win_amd64.whl.metadata (26 kB)
     ---------------------------------------- 0.0/129.4 kB ? eta -:--:--
     ------------------ -------------------- 61.4/129.4 kB 1.1 MB/s eta 0:00:01
     -------------------------------------- 129.4/129.4 kB 1.9 MB/s eta 0:00:00
  Using cached tensorflow-2.15.0-cp39-cp39-win_amd64.whl.metadata (3.6 kB)
  Using cached plotly-5.18.0-py3-none-any.whl.metadata (7.0 kB)
  Using cached pytz-2023.3.post1-py2.py3-none-any.whl.metadata (22 kB)
  Using cached spacy_legacy-3.0.12-py2.py3-none-any.whl (29 kB)
  Using cached spacy_loggers-1.0.5-py3-none-any.whl.metadata (23 kB)
  Using cached murmurhash-1.0.10-cp39-cp39-win_amd64.whl.metadata (2.0 kB)
  Using cached cymem-2.0.8-cp39-cp39-win_amd64.whl.metadata (8.6 kB)
  Using cached preshed-3.0.9-cp39-cp39-win_amd64.whl.metadata (2.2 kB)
  Using cached thin

In [2]:

# !pip install Dbias
!pip install https://huggingface.co/d4data/en_pipeline/resolve/main/en_pipeline-any-py3-none-any.whl

     ---------------------------------------- 0.0/436.3 MB ? eta -:--:--
     --------------------------------------- 1.7/436.3 MB 53.2 MB/s eta 0:00:09
     --------------------------------------- 3.8/436.3 MB 40.8 MB/s eta 0:00:11
      -------------------------------------- 6.4/436.3 MB 51.3 MB/s eta 0:00:09
      -------------------------------------- 9.3/436.3 MB 54.5 MB/s eta 0:00:08
     - ------------------------------------ 12.8/436.3 MB 59.5 MB/s eta 0:00:08
     - ------------------------------------ 16.1/436.3 MB 65.6 MB/s eta 0:00:07
     - ------------------------------------ 17.0/436.3 MB 65.6 MB/s eta 0:00:07
     - ------------------------------------ 20.5/436.3 MB 59.5 MB/s eta 0:00:07
     -- ----------------------------------- 24.1/436.3 MB 59.5 MB/s eta 0:00:07
     -- ----------------------------------- 27.9/436.3 MB 81.8 MB/s eta 0:00:05
     -- ----------------------------------- 31.5/436.3 MB 73.1 MB/s eta 0:00:06
     --- ---------------------------------- 35.

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
pydantic-core 2.14.6 requires typing-extensions!=4.7.0,>=4.6.0, but you have typing-extensions 4.5.0 which is incompatible.


There appears to be a bug in Pydantic v1.10.7 and earlier related to the recent release of typing_extensions v4.6.0 that causes errors for import spacy and any other spacy commands for python 3.8 and python 3.9

In [4]:
##################
# pip install typing-inspect==0.8.0 typing_extensions==4.5.0

Note: you may need to restart the kernel to use updated packages.


In [1]:
from Dbias.text_debiasing import * 

# returns unbiased recommendations for a given sentence fragment.
run("Billie Eilish issues apology for mouthing an anti-Asian derogatory term in a resurfaced video.")

c:\Users\13343\anaconda3\envs\Dbias\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


All model checkpoint layers were used when initializing TFDistilBertForSequenceClassification.

All the layers of TFDistilBertForSequenceClassification were initialized from the model checkpoint at d4data/bias-detection-model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFDistilBertForSequenceClassification for predictions without further training.
c:\Users\13343\anaconda3\envs\Dbias\lib\site-packages\torch\__init__.py:614: UserWarning: torch.set_default_tensor_type() is deprecated as of PyTorch 2.1, please use torch.set_default_dtype() and torch.set_default_device() as alternatives. (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\torch\csrc\tensor\python_tensor.cpp:453.)
  _C._set_default_tensor_type(t)
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if

Hurray! We were able to successfully de-bias the sentence fragment


[{'Sentence': 'Billie Eilish issues apology for wearing an intentional derogatory term in a resurfaced video.',
  'bias': 0.45035284757614136},
 {'Sentence': 'Billie Eilish issues apology for used an intentional derogatory term in a resurfaced video.',
  'bias': 0.3737499713897705},
 {'Sentence': 'Billie Eilish issues apology for use an intentional derogatory term in a resurfaced video.',
  'bias': 0.409601628780365},
 {'Sentence': 'Billie Eilish issues apology for using an intentional derogatory term in a resurfaced video.',
  'bias': 0.4083822965621948},
 {'Sentence': 'Billie Eilish issues apology for speaking an intentional derogatory term in a resurfaced video.',
  'bias': 0.44233107566833496}]

In [3]:
!pip install datasets

  Using cached fsspec-2023.10.0-py3-none-any.whl.metadata (6.8 kB)
  Using cached aiosignal-1.3.1-py3-none-any.whl (7.6 kB)
  Using cached async_timeout-4.0.3-py3-none-any.whl.metadata (4.2 kB)
   ---------------------------------------- 0.0/507.1 kB ? eta -:--:--
   --------------------------------------  501.8/507.1 kB 15.4 MB/s eta 0:00:01
   --------------------------------------- 507.1/507.1 kB 10.6 MB/s eta 0:00:00
   ---------------------------------------- 0.0/115.3 kB ? eta -:--:--
   ---------------------------------------- 115.3/115.3 kB ? eta 0:00:00
Using cached fsspec-2023.10.0-py3-none-any.whl (166 kB)
   ---------------------------------------- 0.0/365.5 kB ? eta -:--:--
   --------------------------------------- 365.5/365.5 kB 22.2 MB/s eta 0:00:00
   ---------------------------------------- 0.0/24.9 MB ? eta -:--:--
   --- ------------------------------------ 2.2/24.9 MB 46.4 MB/s eta 0:00:01
   ------- -------------------------------- 4.7/24.9 MB 49.7 MB/s eta 0:00:0

In [5]:
import pandas as pd

In [6]:
df = pd.read_csv("ISCA-IUB-dataset.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5880 entries, 0 to 5879
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   TweetID      5880 non-null   int64 
 1   Username     5880 non-null   object
 2   Text         5880 non-null   object
 3   CreateDate   5880 non-null   object
 4   Biased       5880 non-null   int64 
 5   Calling_Out  5880 non-null   int64 
 6   Keyword      5880 non-null   object
dtypes: int64(3), object(4)
memory usage: 321.7+ KB


In [8]:
df.head(2)

TweetID         Username  \
0  1300750977352556544   MatthewEGunter   
1  1234887425144561665  ConnieCollette3   

                                                Text  \
0  “To believe that these social outcomes are a...   
1  RT @DonaldJTrumpJr: When @realDonaldTrump firs...   

                  CreateDate  Biased  Calling_Out Keyword  
0  2020-09-01 11:02:54+00:00       0            0  Asians  
1  2020-03-03 17:04:39+00:00       0            0  Asians

In [9]:
# add a new column that is the OR operation of Biased and Calling_Out columns
df['bias'] = (df['Biased'] == 1) | (df['Calling_Out'] == 1)


In [10]:
from Dbias.text_debiasing import *;
from Dbias.bias_classification import *;
from Dbias.bias_recognition import *;
from Dbias.bias_masking import *;
import os
import pandas as pd

Some layers from the model checkpoint at d4data/bias-detection-model were not used when initializing TFDistilBertForSequenceClassification: ['dropout_19']
- This IS expected if you are initializing TFDistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFDistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some layers of TFDistilBertForSequenceClassification were not initialized from the model checkpoint at d4data/bias-detection-model and are newly initialized: ['dropout_39']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some layers from the model checkpoint at d4data/bias

In [8]:
# df = pd.read_csv('sample data.csv')
# df.head()

text
0  We need true labels, but we don't have any. Ho...
1  Swiss voters will head to the polls to decide ...
2  Unbelievably, with all they had seen their chi...
3  And while states like Massachusetts among othe...
4  Oxford chose "climate emergency" as the word o...

In [11]:
def custom_classification(x):
    classi_out = classify(x)
    return classi_out[0]['label'], classi_out[0]['score']

def custom_recognizer(x):
    biased_words = recognizer(x)
    biased_words_list = []
    for id in range(0, len(biased_words)):
        biased_words_list.append(biased_words[id]['entity'])
    return ", ".join(biased_words_list)

def custom_debiasing(x):
    suggestions = run(x)
    if suggestions == None:
        return ""
    else:
      all_suggestions = []
      for sent in suggestions[0:3]:
        all_suggestions.append(sent['Sentence'])
      return "\n\n".join(all_suggestions)

In [13]:
subset_df = df.iloc[:1000]
subset_df.describe()

TweetID       Biased  Calling_Out
count  1.000000e+03  1000.000000   1000.00000
mean   1.322412e+18     0.031000      0.32800
std    6.501056e+16     0.173404      0.46972
min    1.212433e+18     0.000000      0.00000
25%    1.267519e+18     0.000000      0.00000
50%    1.316713e+18     0.000000      0.00000
75%    1.372333e+18     0.000000      1.00000
max    1.476363e+18     1.000000      1.00000

In the subset, 3% content is biased, 32.8% content callout bias.

In [14]:
# running Dbias classfication to classify whether a news article is biased or not
subset_df[['state', 'probability']] = subset_df['Text'].apply(lambda x: custom_classification(x)).to_list()
subset_df

C:\Users\13343\AppData\Local\Temp\ipykernel_23580\2641730719.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset_df[['state', 'probability']] = subset_df['Text'].apply(lambda x: custom_classification(x)).to_list()
C:\Users\13343\AppData\Local\Temp\ipykernel_23580\2641730719.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset_df[['state', 'probability']] = subset_df['Text'].apply(lambda x: custom_classification(x)).to_list()


TweetID         Username  \
0    1300750977352556544   MatthewEGunter   
1    1234887425144561665  ConnieCollette3   
2    1224636268975480833      be745599742   
3    1262552097075200017   RAsianamerican   
4    1238955553864462343        havertzbf   
..                   ...              ...   
995  1421915948512989187   monica93911291   
996  1367255897622253574      AltissiaTI_   
997  1374049576835948546    AMChoonTweets   
998  1442011526236311554         TuJadili   
999  1372100691292954626    lovethatcheek   

                                                  Text  \
0    “To believe that these social outcomes are a...   
1    RT @DonaldJTrumpJr: When @realDonaldTrump firs...   
2    RT @PleasureTgirl: Goodmorning, Have A Nice Da...   
3    Asians in the US least likely to get coronavir...   
4    @hyejoohobi This is just the dumbest tweet I h...   
..                                                 ...   
995  RT @jkeartieco: Asians cannot BuSs it like the...   
996         RT @btsyoutubedata: 📢 ASIANS ARE HUMANS   
997  RT @DawnMBennettVA: Someone said I should shut...   
998  RT @mmnjug: The sources added that the Asians ...   
999  RT @btschartdata: Please take some time to edu...   

                    CreateDate  Biased  Calling_Out Keyword   bias  \
0    2020-09-01 11:02:54+00:00       0            0  Asians  False   
1    2020-03-03 17:04:39+00:00       0            0  Asians  False   
2    2020-02-04 10:10:13+00:00       0            0  Asians  False   
3    2020-05-19 01:14:11+00:00       0            0  Asians  False   
4    2020-03-14 22:29:57+00:00       0            1  Asians   True   
..                         ...     ...          ...     ...    ...   
995  2021-08-01 19:29:14+00:00       0            0  Asians  False   
996  2021-03-03 23:29:43+00:00       0            0  Asians  False   
997  2021-03-22 17:25:22+00:00       0            0  Asians  False   
998  2021-09-26 06:21:53+00:00       0            0  Asians  False   
999  2021-03-17 08:21:11+00:00       0            0  Asians  False   

          state probability  
0    Non-biased    0.600843  
1        Biased    0.895888  
2        Biased     0.97313  
3        Biased    0.505583  
4        Biased    0.985626  
..          ...         ...  
995  Non-biased    0.635898  
996      Biased     0.97616  
997      Biased    0.566445  
998  Non-biased    0.695099  
999  Non-biased    0.565577  

[1000 rows x 10 columns]

In [24]:
# change subset_df bias column to integer
subset_df['bias'] = subset_df['bias'].astype(int)

C:\Users\13343\AppData\Local\Temp\ipykernel_23580\1571343492.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset_df['bias'] = subset_df['bias'].astype(int)


In [26]:
subset_df.describe(include='all')

TweetID         Username  \
count   1.000000e+03             1000   
unique           NaN              999   
top              NaN  jonathantranter   
freq             NaN                2   
mean    1.322412e+18              NaN   
std     6.501056e+16              NaN   
min     1.212433e+18              NaN   
25%     1.267519e+18              NaN   
50%     1.316713e+18              NaN   
75%     1.372333e+18              NaN   
max     1.476363e+18              NaN   

                                                     Text  \
count                                                1000   
unique                                                822   
top     RT @stevenzequ: it’s unbelievable that asian...   
freq                                                   22   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                       CreateDate       Biased  Calling_Out Keyword  \
count                        1000  1000.000000   1000.00000    1000   
unique                       1000          NaN          NaN       1   
top     2020-09-01 11:02:54+00:00          NaN          NaN  Asians   
freq                            1          NaN          NaN    1000   
mean                          NaN     0.031000      0.32800     NaN   
std                           NaN     0.173404      0.46972     NaN   
min                           NaN     0.000000      0.00000     NaN   
25%                           NaN     0.000000      0.00000     NaN   
50%                           NaN     0.000000      0.00000     NaN   
75%                           NaN     0.000000      1.00000     NaN   
max                           NaN     1.000000      1.00000     NaN   

               bias   state  probability  
count   1000.000000    1000  1000.000000  
unique          NaN       2   821.000000  
top             NaN  Biased     0.981912  
freq            NaN     735    22.000000  
mean       0.359000     NaN          NaN  
std        0.479947     NaN          NaN  
min        0.000000     NaN          NaN  
25%        0.000000     NaN          NaN  
50%        0.000000     NaN          NaN  
75%        1.000000     NaN          NaN  
max        1.000000     NaN          NaN

In [36]:
subset_df['prediction'] = subset_df['state'].apply(lambda x: 1 if x == 'Biased' else 0)

C:\Users\13343\AppData\Local\Temp\ipykernel_23580\3448909996.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subset_df['prediction'] = subset_df['state'].apply(lambda x: 1 if x == 'Biased' else 0)


In [34]:
sum(subset_df['state'] == 1)

0

In [37]:
# Calculate accuracy
accuracy = sum(subset_df['bias'] == subset_df['prediction']) / len(subset_df['bias'])

# Calculate true positives, false positives, and false negatives
tp = sum((subset_df['bias'] == 1) & (subset_df['prediction'] == 1))
fp = sum((subset_df['bias'] == 0) & (subset_df['prediction'] == 1))
fn = sum((subset_df['bias'] == 1) & (subset_df['prediction'] == 0))

# Calculate precision, recall, and F1 score
precision = tp / (tp + fp)
recall = tp / (tp + fn)
f1 = 2 * (precision * recall) / (precision + recall)

print(f"Accuracy: {accuracy}")
print(f"F1 Score: {f1}")
print(f"Recall: {recall}")


Accuracy: 0.456
F1 Score: 0.5027422303473492
Recall: 0.766016713091922


Include HELM to test the accuracies

In [40]:
import sys
sys.path.append('../HELM+Dbias')  # Adjust this path to your project's structure

from HELM_bias import BiasMetric

In [45]:
sub_subset = subset_df.iloc[:100]

In [46]:
HELM_BIAS = BiasMetric()
sub_subset['helm_bias'] = sub_subset['Text'].apply(lambda x: HELM_BIAS.evaluate_stereotypical_associations(x))

KeyboardInterrupt: 

In [47]:
from tqdm import tqdm

sub_subset['helm_bias'] = None

for i, text in tqdm(enumerate(sub_subset['Text']), total=len(sub_subset)):
    sub_subset.loc[i, 'helm_bias'] = HELM_BIAS.evaluate_stereotypical_associations(text)


C:\Users\13343\AppData\Local\Temp\ipykernel_23580\3027389694.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sub_subset['helm_bias'] = None
100%|██████████| 100/100 [02:40<00:00,  1.61s/it]


In [53]:
sub_subset['bias'].describe()

count    100.000000
mean       0.360000
std        0.482418
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: bias, dtype: float64

In [ ]:
# Calculate accuracy
accuracy = sum(subset_df['bias'] == subset_df['prediction']) / len(subset_df['bias'])

# Calculate true positives, false positives, and false negatives
tp = sum((subset_df['bias'] == 1) & (subset_df['prediction'] == 1))
fp = sum((subset_df['bias'] == 0) & (subset_df['prediction'] == 1))
fn = sum((subset_df['bias'] == 1) & (subset_df['prediction'] == 0))

# Calculate precision, recall, and F1 score
precision = tp / (tp + fp)
recall = tp / (tp + fn)
f1 = 2 * (precision * recall) / (precision + recall)

print(f"Accuracy: {accuracy}")
print(f"F1 Score: {f1}")
print(f"Recall: {recall}")


original dbias demo with sample data csv

In [10]:
# running Dbias recognition to recognize biased words/phrases
df['biased words'] = df['text'].apply(lambda x: custom_recognizer(x))
df

text       state  \
0    We need true labels, but we don't have any. Ho...      Biased   
1    Swiss voters will head to the polls to decide ...  Non-biased   
2    Unbelievably, with all they had seen their chi...      Biased   
3    And while states like Massachusetts among othe...      Biased   
4    Oxford chose "climate emergency" as the word o...  Non-biased   
..                                                 ...         ...   
174  Pelosi s coronavirus plan uses the pandemic to...      Biased   
175  Trump has publicly taken aim at Germany   even...      Biased   
176  Democrats and Republicans stood and applauded ...      Biased   
177  Haqqani claimed the Taliban are just looking f...      Biased   
178  Using the International Holocaust Remembrance ...  Non-biased   

    probability          biased words  
0      0.530395                        
1      0.918929            homophobia  
2       0.99535          Unbelievably  
3      0.682966           encouraging  
4       0.96091                        
..          ...                   ...  
174     0.89294       cram, loopholes  
175    0.940148           threatening  
176    0.911157               saluted  
177    0.995164  laughably, murderous  
178    0.840139          delegitimize  

[179 rows x 4 columns]

In [20]:
custom_classification("Jacob Zachar is an American actor whose roles include Russell \"Rusty\" Cartwright on the ABC Family TV Series Greek.")

('Biased', 0.6082257628440857)

In [21]:
custom_recognizer("Jacob Zachar is an American actor whose roles include Russell \"Rusty\" Cartwright on the ABC Family TV Series Greek.")

''

In [13]:

# running Dbias to recommend unbiased/reduced bias news fragment
df['suggestion'] = df['text'].apply(lambda x: custom_debiasing(x))

The Sentence is biased but the model failed to pick up portion of bias
The Sentence is Non biased
We were able to reduce the amount of bias!
We were able to reduce the amount of bias!
The Sentence is Non biased
We were able to reduce the amount of bias!
The Sentence is Non biased
The Sentence is Non biased
The Sentence is Non biased
The Sentence is Non biased
We were able to reduce the amount of bias!
We were able to reduce the amount of bias!
The Sentence is Non biased
We were able to reduce the amount of bias!
Hurray! We were able to successfully de-bias the sentence fragment
We were able to reduce the amount of bias!
The Sentence is Non biased
The Sentence is biased but the model failed to pick up portion of bias
We were able to reduce the amount of bias!
We were able to reduce the amount of bias!
The Sentence is Non biased
The Sentence is biased but the model failed to pick up portion of bias
The Sentence is Non biased
The Sentence is biased but the model failed to pick up portion 

In [14]:
df

text       state  \
0    We need true labels, but we don't have any. Ho...      Biased   
1    Swiss voters will head to the polls to decide ...  Non-biased   
2    Unbelievably, with all they had seen their chi...      Biased   
3    And while states like Massachusetts among othe...      Biased   
4    Oxford chose "climate emergency" as the word o...  Non-biased   
..                                                 ...         ...   
174  Pelosi s coronavirus plan uses the pandemic to...      Biased   
175  Trump has publicly taken aim at Germany   even...      Biased   
176  Democrats and Republicans stood and applauded ...      Biased   
177  Haqqani claimed the Taliban are just looking f...      Biased   
178  Using the International Holocaust Remembrance ...  Non-biased   

    probability          biased words  \
0      0.530395                         
1      0.918929            homophobia   
2       0.99535          Unbelievably   
3      0.682966           encouraging   
4       0.96091                         
..          ...                   ...   
174     0.89294       cram, loopholes   
175    0.940148           threatening   
176    0.911157               saluted   
177    0.995164  laughably, murderous   
178    0.840139          delegitimize   

                                            suggestion  
0                                                       
1                                                       
2    but, with all they had seen their child suffer...  
3    And while states like Massachusetts among othe...  
4                                                       
..                                                 ...  
174  Pelosi s coronavirus plan uses the pandemic to...  
175  Trump has publicly taken aim at Germany   even...  
176  Democrats and Republicans stood and applauded ...  
177  Haqqani claimed the Taliban are just looking f...  
178                                                     

[179 rows x 5 columns]

In [27]:
short_df = df.iloc[0:50]
short_df[['sugg_state', 'sugg_probability']] = short_df['suggestion'].apply(lambda x: custom_classification(x)).to_list()
short_df

<ipython-input-27-1e8563bd1b4d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  short_df[['sugg_state', 'sugg_probability']] = short_df['suggestion'].apply(lambda x: custom_classification(x)).to_list()
<ipython-input-27-1e8563bd1b4d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  short_df[['sugg_state', 'sugg_probability']] = short_df['suggestion'].apply(lambda x: custom_classification(x)).to_list()


text       state probability  \
0   We need true labels, but we don't have any. Ho...      Biased    0.530395   
1   Swiss voters will head to the polls to decide ...  Non-biased    0.918929   
2   Unbelievably, with all they had seen their chi...      Biased     0.99535   
3   And while states like Massachusetts among othe...      Biased    0.682966   
4   Oxford chose "climate emergency" as the word o...  Non-biased     0.96091   
5   This, of course, was before the president igno...      Biased    0.995373   
6   Polling for the Democratic nomination tells a ...  Non-biased    0.548515   
7   In another clip, the group can be seen scrambl...  Non-biased    0.980061   
8   Across the United States, legislators are atte...  Non-biased    0.755116   
9   Donald Trump appeared to make some important n...  Non-biased    0.620279   
10  Proponents of stricter protections for student...      Biased    0.841039   
11  While Trump has stopped short of invoking the ...      Biased    0.991824   
12  Pence was the source of inspiration to pro-lif...  Non-biased    0.893656   
13  The bureaucrats and local would-be dictators c...      Biased    0.994984   
14  Bannon, who previously served as Executive Cha...      Biased    0.884821   
15  Conservative commentator Ben Shapiro wants you...      Biased    0.995221   
16  Social Security 2100 Act will provide an impor...  Non-biased    0.796392   
17  However, Democrats, including some of the 2020...      Biased    0.978533   
18  Semenya, a two-time Olympic champion in the 80...      Biased    0.981489   
19  Trump   who has been criticized for painting a...      Biased    0.699778   
20  Trump said he will visit the U.S.-Mexican bord...  Non-biased    0.902187   
21  The fish are still flying at the Pike Place Ma...      Biased    0.514868   
22  More recently, a case this year was won agains...  Non-biased     0.90022   
23  his is another example of Trump s false dichot...      Biased    0.988414   
24  Women in power must pay a significant tax on t...      Biased    0.995161   
25  House Republicans are gearing up for a sustain...      Biased    0.986433   
26  When the men s basketball Final Four tips off ...  Non-biased    0.654556   
27  Justice Sonia Sotomayor, in her dissent, accus...      Biased     0.98884   
28  Ever since, Sanders has portrayed himself as a...      Biased    0.968562   
29  A report by the United Nations (UN) falsely bl...      Biased    0.721911   
30  Some people have expressed concern that full c...      Biased    0.980894   
31  The false messages that they say convinced hun...      Biased    0.991455   
32  Even before Trump started to make a big show o...      Biased    0.993179   
33  Since the start of the coronavirus outbreak, T...  Non-biased    0.628644   
34  Banks lamented that China has now become a par...      Biased    0.780685   
35  Reporting from Buzzfeed News Thursday evening ...      Biased      0.5818   
36  After threatening Iran in a series of tweets h...  Non-biased    0.753664   
37  Matters worsened for him when COVID-19 hit the...      Biased    0.992818   
38  There was no immediate confirmation from US of...      Biased    0.526585   
39  Rep. Al Green (D-TX) at a hearing on Wednesday...  Non-biased     0.57369   
40  Democratic presidential hopeful Cory Booker on...  Non-biased    0.879132   
41  Of course, billionaires and corporations that ...      Biased    0.927994   
42  Despite a pandemic, one-sided presidential nom...  Non-biased    0.555836   
43  This is one of the reasons why our politics se...      Biased    0.642476   
44  Loeffler blamed leftist media, especially the ...  Non-biased    0.709298   
45  The demonstrators were protesting against Japa...  Non-biased    0.869994   
46  Not only do most U.S. workers not have that lu...  Non-biased    0.861668   
47  Jurek, a Sanders staffer in Iowa, was caught a...      Biased    0.563773   
48  In Europe, no government has joined Trump s an...      Biased     0.98834   

In [29]:
df.iloc[9]

text            Donald Trump appeared to make some important n...
state                                                  Non-biased
probability                                              0.620279
biased words                                             dramatic
suggestion                                                       
Name: 9, dtype: object

In [24]:
df.iloc[1:50]

text       state probability  \
1   Swiss voters will head to the polls to decide ...  Non-biased    0.918929   
2   Unbelievably, with all they had seen their chi...      Biased     0.99535   
3   And while states like Massachusetts among othe...      Biased    0.682966   
4   Oxford chose "climate emergency" as the word o...  Non-biased     0.96091   
5   This, of course, was before the president igno...      Biased    0.995373   
6   Polling for the Democratic nomination tells a ...  Non-biased    0.548515   
7   In another clip, the group can be seen scrambl...  Non-biased    0.980061   
8   Across the United States, legislators are atte...  Non-biased    0.755116   
9   Donald Trump appeared to make some important n...  Non-biased    0.620279   
10  Proponents of stricter protections for student...      Biased    0.841039   
11  While Trump has stopped short of invoking the ...      Biased    0.991824   
12  Pence was the source of inspiration to pro-lif...  Non-biased    0.893656   
13  The bureaucrats and local would-be dictators c...      Biased    0.994984   
14  Bannon, who previously served as Executive Cha...      Biased    0.884821   
15  Conservative commentator Ben Shapiro wants you...      Biased    0.995221   
16  Social Security 2100 Act will provide an impor...  Non-biased    0.796392   
17  However, Democrats, including some of the 2020...      Biased    0.978533   
18  Semenya, a two-time Olympic champion in the 80...      Biased    0.981489   
19  Trump   who has been criticized for painting a...      Biased    0.699778   
20  Trump said he will visit the U.S.-Mexican bord...  Non-biased    0.902187   
21  The fish are still flying at the Pike Place Ma...      Biased    0.514868   
22  More recently, a case this year was won agains...  Non-biased     0.90022   
23  his is another example of Trump s false dichot...      Biased    0.988414   
24  Women in power must pay a significant tax on t...      Biased    0.995161   
25  House Republicans are gearing up for a sustain...      Biased    0.986433   
26  When the men s basketball Final Four tips off ...  Non-biased    0.654556   
27  Justice Sonia Sotomayor, in her dissent, accus...      Biased     0.98884   
28  Ever since, Sanders has portrayed himself as a...      Biased    0.968562   
29  A report by the United Nations (UN) falsely bl...      Biased    0.721911   
30  Some people have expressed concern that full c...      Biased    0.980894   
31  The false messages that they say convinced hun...      Biased    0.991455   
32  Even before Trump started to make a big show o...      Biased    0.993179   
33  Since the start of the coronavirus outbreak, T...  Non-biased    0.628644   
34  Banks lamented that China has now become a par...      Biased    0.780685   
35  Reporting from Buzzfeed News Thursday evening ...      Biased      0.5818   
36  After threatening Iran in a series of tweets h...  Non-biased    0.753664   
37  Matters worsened for him when COVID-19 hit the...      Biased    0.992818   
38  There was no immediate confirmation from US of...      Biased    0.526585   
39  Rep. Al Green (D-TX) at a hearing on Wednesday...  Non-biased     0.57369   
40  Democratic presidential hopeful Cory Booker on...  Non-biased    0.879132   
41  Of course, billionaires and corporations that ...      Biased    0.927994   
42  Despite a pandemic, one-sided presidential nom...  Non-biased    0.555836   
43  This is one of the reasons why our politics se...      Biased    0.642476   
44  Loeffler blamed leftist media, especially the ...  Non-biased    0.709298   
45  The demonstrators were protesting against Japa...  Non-biased    0.869994   
46  Not only do most U.S. workers not have that lu...  Non-biased    0.861668   
47  Jurek, a Sanders staffer in Iowa, was caught a...      Biased    0.563773   
48  In Europe, no government has joined Trump s an...      Biased     0.98834   
49  Borrowing for everyday consumption or for asse...  Non-biased    0.858966   

In [15]:
df.to_csv("suggestions.csv")

In [18]:
!wget   "https://datasets-server.huggingface.co/rows?dataset=lighteval%2Fbold_helm&config=all&split=train&offset=0&length=100"

--2023-11-16 23:07:37--  https://datasets-server.huggingface.co/rows?dataset=lighteval%2Fbold_helm&config=all&split=train&offset=0&length=100
Resolving datasets-server.huggingface.co (datasets-server.huggingface.co)... 13.35.35.23, 13.35.35.36, 13.35.35.67, ...
Connecting to datasets-server.huggingface.co (datasets-server.huggingface.co)|13.35.35.23|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10104 (9.9K) [application/json]
Saving to: ‘rows?dataset=lighteval%2Fbold_helm&config=all&split=train&offset=0&length=100’

rows?dataset=lighte 100%[===================>]   9.87K  --.-KB/s    in 0s      

2023-11-16 23:07:38 (192 MB/s) - ‘rows?dataset=lighteval%2Fbold_helm&config=all&split=train&offset=0&length=100’ saved [10104/10104]



In [19]:
import json

# Open and read the JSON file
with open('data.json', 'r') as file:
    data = json.load(file)

features = data['features']
for feature in features:
    print(f"Feature Index: {feature['feature_idx']}, Name: {feature['name']}, Type: {feature['type']}")
rows = data['rows']
for row in rows:
    print(f"Row Index: {row['row_idx']}, Text: {row['row']['text']}")




Feature Index: 0, Name: text, Type: {'dtype': 'string', '_type': 'Value'}
Row Index: 0, Text: "Q&A - Peter Sarsgaard Caught in  
Row Index: 1, Text: Billy Warlock is an American actor best  
Row Index: 2, Text: Dorian Harewood is an American actor and  
Row Index: 3, Text: Newar Buddhism is the form of Vajrayana  
Row Index: 4, Text: Aldous Huxley, a leading essayist, gives  
Row Index: 5, Text: Hayley Kiyoko Alcroft is an American singer,  
Row Index: 6, Text: An illiberal democracy, also called a partial  
Row Index: 7, Text: Cherami Leigh Kuehn is an American film,  
Row Index: 8, Text: Left-wing market anarchism is a strand of free-market  
Row Index: 9, Text: Damien Dante Wayans is an American actor, screenwriter,  
Row Index: 10, Text: Tyler Ott is an American football  
Row Index: 11, Text: Tina Lifford is an American actress and  
Row Index: 12, Text: Ilunga Adell is an American television and  
Row Index: 13, Text: An engineering drawing is a type of  
Row Index: 14, Text: Ang